### ISSUE TITLE
Parallelisation of the conservative bounding

### ISSUE THREAD

**Issue by mihailovnik** [isoAdvector/isoAdvector#11]
Hello 

At present moment a cell will be checked for bounding if this is a surface cell or a neighbour of surface cell via internal face or a neighbour of neighbour of surface cell via internal faces. Hence in the case of a surface cell which is border cell for a processor patch the corresponding neighbour cells in the neighbour processor will not be checked. Therefore the parallel computation may give another result than serial one.

I am right?

**Comment by isoAdvector** [comment]
Yes. Generally with OpenFOAM simulations, you should not expect to get the exact same results down to machine precision in parallel and serial. However, since isoAdvector is explicit in nature, this could be fixed fairly easily to give exact correspondence between serial and parallel runs. Care must however be taken in limiting the number of times information is transferred between subdomains since there is an overhead associated with each call up. Not sure how critical this is though.

**Comment by mihailovnik** [comment]
Thanks for answer!
Most likely the exact correspondence between serial and parallel runs will demand additional information exchange between subdomains. But I assume there are cases when only one cell which can return or accept "excess" volume is on neighbour subdomain. Hence conservative bounding will be unsuccessful. But it is only the assumption. There is a sense to try to think up the example.

**Comment by ma-tri-x** [comment]
Hello isoAdvector programmers!
I'm don't know whether this issue addresses my difficulties, but I'm going to ask the question here:
I'm using foam-extend-3.2 and successfully incorporated isoAdvector. We even were able to modify it for compressible two-phase-flow. However: even with the incompressible formulation (interFlow), when a "bubble" passes a processor patch, clear miscalculations in the alpha1-field can be seen in the cells that are members of the processor-patch. They don't "snap" back to alpha1=1, once they have been lower. 
Secondly, I wonder why it is impossible to use isoAdvector for a 1D-mesh. In the first timestep, thousands of foam warnings of type
```
--> FOAM Warning : 
    From function Foam::scalar Foam::isoAdvection::timeIntegratedFlux(...)
    in file isoAdvection/isoAdvection.C at line 528
    Warning: nShifts = 0 on face 1974 with pTimes = 4{5.1200006e-12} owned by cell 1974
```
arise. Even for single core calculation.
Thirdly, I think there must be some kind of bug in the quadAreaCoeffs function: We've experienced some floating point issues where the coeffs B and A are calculated. I've added one more condition:
FROM
```
if (np0 > 0 && np1 > 0 )
    {
        // Defining quadrilateral vertices
        vector A(pf0[0]);
        vector C(pf1[0]);
        vector B(vector::zero);
        vector D(vector::zero);
       ...
```
TO
```
if (np0 > 0 && np1 > 0 && !(np0 == 1 && np1 ==1)) //MK: added "&& !(np0 == 1 && np1 ==1)"
    {
        // Defining quadrilateral vertices
        vector A(pf0[0]);
        vector C(pf1[0]);
        vector B(vector::zero);
        vector D(vector::zero);
       ...
```
Because B sometimes was filled with some numbers like 1e234 when both pf's were of size=1. It seems to me, correct me if I'm wrong, that this happens, when the isoFace is exactly parallel to the cell-face it advects to.

Am I the only one experiencing this?
Best regards and thank you so much for inventing IsoAdvector!

**Comment by isoAdvector** [comment]
Hello ma-tri-x

Thanks a lot for your feedback and for not only using but also extending the applicability of isoAdvector. Are you planning to publish your compressible solver at some point? What is your target application area?

Regarding your questions, it is not clear to me whether you are using the isoAdvector in the OpenFOAM directory or the one in the foam-extend directory. Based on your lines from quadAreaCoeffs, it looks like you are using the foam-extend version. I'd strongly advice you to base your development on the OpenFOAM version, since I have not touched the foam-extend version for a long time and do not have resources to update it in the near future. The OpenFOAM version contains some very important bug fixes that probably cause some of the problems you experience. Note that if you have e.g. OpenFOAM-2.2.0 sourced, the script bin/adjustCodeToLoadedOFVersion will modify the isoAdvector code reversing the API changes done in the latest OpenFOAM versions and put the resulting code in an OpenFOAM-2.2.0 directory (at least if you run it via the Allwmake script). The resulting isoAdvector code will probably be much closer to the foam-extend-3.2 API (I am guessing here), so you will only have to do much fewer changes manually to get it to work.

In fact here's a challenge: Modify the bin/adjustCodeToLoadedOFVersion script so it makes the resulting isoAdvector version compile without errors with foam-extend sourced. This would be very useful to a lot of people!

Regarding 1D meshes: I did not work with this myself but I don't see why it should not work. Does it crash or only give you warnings? The warnings might be caused by an initial interface exactly coinciding with the mesh faces (like if you have used setFields to initialise alpha1). Maybe this will also be solved by migrating to the OpenFOAM isoAdvector version (rather than the foam-extend version). Same for your parallelisation issue.

Kind regards,
Johan

Ps: Please start a new issue instead of commenting on an existing issue, if your comment is not strictly related to that issue.
